uplc-builtins.md

```diff
@@ -16,22 +16,18 @@ module UPLC-BUILTINS
   imports UPLC-STRING-BUILTINS
   imports UPLC-DATA-BUILTINS
 
-  rule <k> (builtin BN) => < builtin BN .List size(#typeSignature(BN)) > ... </k>
-       <env> _ => .Map </env>
-    requires notBool isPolyBuiltinName(BN)
-     andBool notBool BN ==K chooseData
+  syntax Bool ::= Value        "~" TypeVariable [function, klabel(typeCompatible),    symbol]
+                | TypeConstant "~" TypeVariable [function, klabel(typeCompatibleAux), symbol]
+  rule < con A _ > ~ X                       => A ~ X
+  rule _:Value     ~ _:FullyPolyTypeVariable => true
+  rule _:Value     ~ _                       => false [owise]
 
-  rule <k> (builtin (_::PolyBuiltinName #Or chooseData) #as BN) ~> Force => < builtin BN .List size(#typeSignature(BN)) > ... </k>
-       <env> _ => .Map </env>
-
-  rule #typeCheck(ListItem(< con T _ >)          L, ListItem(T:TypeConstant) TS:List) => #typeCheck(L, TS)
-  rule #typeCheck(ListItem(_:Value)              L, ListItem(anyValue)       TS:List) => #typeCheck(L, TS)
-  rule #typeCheck(ListItem(< con list(_) _ >)    L, ListItem(anyList)        TS:List) => #typeCheck(L, TS)
-  rule #typeCheck(ListItem(< con pair(_)(_) _ >) L, ListItem(anyPair)        TS:List) => #typeCheck(L, TS)
-  rule #typeCheck(ListItem(< con T _ >)          L, ListItem(mkConsCase)    _TS:List) => #typeCheck(L, ListItem(list(T)))
-
-  rule #typeCheck(.List, _) => true
-  rule #typeCheck(    _, _) => false [owise]
+  rule A:TypeConstant ~ A                         => true
+  rule list(A)        ~ listTV(X)                 => A ~ X
+  rule pair(A)(B)     ~ pairTV(X, Y)              => A ~ X andBool B ~ Y
+  rule _:TypeConstant ~ _:PolyBuiltinTypeVariable => true
+  rule _:TypeConstant ~ _:FullyPolyTypeVariable   => true
```

Comment:
@gtrepta maybe the missing case I pointed out before is covered here?

Reply:
Yeah, these rules might benefit from some rework.

The overloaded function is meant for the case where $\iota \in U$ and $V \in C_{\iota}$, but it looks like it ends up handling other cases.